authorization/token/service/token_service.go

```diff
@@ -0,0 +1,326 @@
+	tokenClaims, err := manager.ParseToken(ctx, tokenString)
+	if err != nil {
+		return nil, errors.NewBadParameterErrorFromString("tokenString", tokenString, "invalid token string could not be parsed")
+	}
+
+	// Now that we have the identity and have parsed the token, we can see if we have a record of the token in the database
+	var tokenID uuid.UUID
+
+	// Extract the kid from the token
+	tokenID, err = uuid.FromString(tokenClaims.Id)
+	if err != nil {
+		return nil, errors.NewBadParameterErrorFromString("jti", tokenClaims.Id, "invalid jti identifier - not a UUID")
+	}
+
+	loadedToken, err := s.Repositories().TokenRepository().Load(ctx, tokenID)
+	if err != nil {
+		// This is not an error per se, so we'll just log an informational message
+		log.Info(ctx, map[string]interface{}{
+			"token_id": tokenID,
+		}, "token with specified id not found")
+	}
+
+	// Check whether the resource exists in the token already (only for valid RPT tokens)
+	resourceExistsInToken := false
+	if loadedToken != nil && tokenClaims.Permissions != nil {
+		for _, tokenPermission := range *tokenClaims.Permissions {
+			if *tokenPermission.ResourceSetID == resourceID {
+				resourceExistsInToken = true
+			}
+		}
+	}
+
+	if loadedToken != nil {
+		// Confirm that the token belongs to the current identity
+		if loadedToken.IdentityID != identity.ID {
+			return nil, errors.NewUnauthorizedError("invalid token for identity")
+		}
+
+		// If the token exists and its status is valid, return an empty string
+		if loadedToken.Valid() && resourceExistsInToken {
+			return nil, nil
+		}
+
+		// We now process the various token status codes in order of priority, starting with DEPROVISIONED
+		if loadedToken.HasStatus(tokenPkg.TOKEN_STATUS_DEPROVISIONED) {
+			return nil, errors.NewUnauthorizedErrorWithCode("token deprovisioned", errors.UNAUTHORIZED_CODE_TOKEN_DEPROVISIONED)
+		}
+
+		// If the token has been revoked or the user is logged out, we respond in the same way
+		if loadedToken.HasStatus(tokenPkg.TOKEN_STATUS_REVOKED) || loadedToken.HasStatus(tokenPkg.TOKEN_STATUS_LOGGED_OUT) {
+			return nil, errors.NewUnauthorizedErrorWithCode("token revoked or logged out", errors.UNAUTHORIZED_CODE_TOKEN_REVOKED)
+		}
+
+		// If the token is stale, yet the resource exists in the token
+		// we can re-evaluate its privileges to determine whether they have changed.
+		// If the privileges are unchanged, then reset the token status
+		if loadedToken.HasStatus(tokenPkg.TOKEN_STATUS_STALE) && resourceExistsInToken {
+			// Query for all of the token's privileges
+			privileges, err := s.Repositories().TokenRepository().ListPrivileges(ctx, tokenID)
+			if err != nil {
+				return nil, errors.NewInternalError(ctx, err)
+			}
+
+			scopesChanged := false
+			permissionsChanged := false
+
+			// If the number of permissions in the token is different to what is stored in the database, then
+			// something has definitely changed
+			if len(privileges) != len(*tokenClaims.Permissions) {
```

Comment:
This could happen in the scenario that a resource is deleted, in which case I believe we should be removing the mapping from the token to the privilege cache (the privilege cache record itself we would simply set `deleted_at` via a soft delete).

Reply:
Right, sounds like a trigger scenario?